libminifi/src/utils/LineByLineInputOutputStreamCallback.cpp

```diff
@@ -26,27 +26,36 @@ LineByLineInputOutputStreamCallback::LineByLineInputOutputStreamCallback(Callbac
   : callback_(std::move(callback)) {
 }
 
-int64_t LineByLineInputOutputStreamCallback::operator()(const std::shared_ptr<io::InputStream>& input, const std::shared_ptr<io::OutputStream>& output) {
+io::ReadWriteResult LineByLineInputOutputStreamCallback::operator()(const std::shared_ptr<io::InputStream>& input, const std::shared_ptr<io::OutputStream>& output) {
   gsl_Expects(input);
   gsl_Expects(output);
 
+  io::ReadWriteResult result;
+
   if (int64_t status = readInput(*input); status <= 0) {
-    return status;
+    result.bytes_read = status;
+    return result;
   }
 
-  std::size_t total_bytes_written_ = 0;
+  result.bytes_read = gsl::narrow<int64_t>(input_.size());
+
+  std::size_t total_bytes_written = 0;
   bool is_first_line = true;
   readLine();
   do {
     readLine();
     std::string output_line = callback_(*current_line_, is_first_line, isLastLine());
     const auto bytes_written = output->write(reinterpret_cast<const uint8_t *>(output_line.data()), output_line.size());
-    if (io::isError(bytes_written)) { return -1; }
-    total_bytes_written_ += bytes_written;
+    if (io::isError(bytes_written)) {
+      result.bytes_written = gsl::narrow<int64_t>(bytes_written);
+      return result;
+    }
```

Comment:
did you mean to return the error code as `result.bytes_written`? It used to return -1.

Reply:
Yes, if the bytes_written returns a -1 error value that should be returned, I think that's consistent with other callback results. Do you suggest something else?

Reply:
this way if errors can be returned in both members, it's confusing where to check. Maybe we could return an optional instead, with empty meaning error, since we're not differentiating between different kinds of errors anyway.

Reply:
I think using optional here is a good idea, updated in https://github.com/apache/nifi-minifi-cpp/pull/1909/commits/24c93329eec9e7a505d318804ba9b78589d5c9d3